pkg/ioutils/bytespipe.go

```diff
@@ -0,0 +1,93 @@
+package ioutils
+
+const maxCap = 1e6
+
+// BytesPipe is io.ReadWriter which works similary to pipe(queue).
+// All written data could be read only once. Also BytesPipe trying to adjust
+// internal []byte slice to current needs, so there won't be overgrown buffer
+// after highload peak.
+// BytesPipe isn't goroutine-safe, caller must synchronize it if needed.
+type BytesPipe struct {
+	buf        [][]byte // slice of byte-slices of buffered data
+	lastRead   int      // index in the first slice to a read point
+	bufLen     int      // length of data buffered over the slices
+	bufCap     int      // whole allocated capacity over slices
+	writeIndex int      // index to the buffer where the writes go
```

Comment:
If the growth loop was merged with the copy loop, this variable wouldn't be necessary, because the writes would always go to the last element of `bp.buf`.